node/subsystem/src/util.rs

```diff
@@ -0,0 +1,570 @@
+			job: std::marker::PhantomData,
+		}
+	}
+
+	/// Spawn a new job for this `parent_hash`, with whatever args are appropriate.
+	fn spawn_job(&mut self, parent_hash: Hash, run_args: Job::RunArgs) -> Result<(), Error> {
+		let (to_job_tx, to_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (from_job_tx, from_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (finished_tx, finished) = oneshot::channel();
+
+		let (future, abort_handle) = future::abortable(async move {
+			if let Err(e) = Job::run(parent_hash, run_args, to_job_rx, from_job_tx).await {
+				log::error!(
+					"{}({}) finished with an error {:?}",
+					Job::NAME,
+					parent_hash,
+					e,
+				);
+			}
+		});
+
+		// discard output
+		let future = async move {
+			let _ = future.await;
+			let _ = finished_tx.send(());
+		};
+		self.spawner.spawn(future)?;
+
+		// this handle lets us remove the appropriate receiver from self.outgoing_msgs
+		// when it's time to stop the job.
+		let outgoing_msgs_handle = self.outgoing_msgs.push(from_job_rx);
+
+		let handle = JobHandle {
+			abort_handle,
+			to_job: to_job_tx,
+			finished,
+			outgoing_msgs_handle,
+		};
+
+		self.running.insert(parent_hash, handle);
+
+		Ok(())
+	}
+
+	/// Stop the job associated with this `parent_hash`.
+	pub async fn stop_job(&mut self, parent_hash: Hash) -> Result<(), Error> {
+		match self.running.remove(&parent_hash) {
+			Some(handle) => {
+				Pin::new(&mut self.outgoing_msgs).remove(handle.outgoing_msgs_handle);
+				handle.stop().await;
+				Ok(())
+			}
+			None => Err(Error::JobNotFound(parent_hash)),
+		}
+	}
+
+	/// Send a message to the appropriate job for this `parent_hash`.
+	async fn send_msg(&mut self, parent_hash: Hash, msg: Job::ToJob) -> Result<(), Error> {
+		match self.running.get_mut(&parent_hash) {
+			Some(job) => job.send_msg(msg).await?,
+			None => return Err(Error::JobNotFound(parent_hash)),
+		}
+		Ok(())
+	}
+
+	/// Get the next message from any of the underlying jobs.
+	async fn next(&mut self) -> Option<Job::FromJob> {
+		self.outgoing_msgs.next().await.and_then(|(e, _)| match e {
+			StreamYield::Item(e) => Some(e),
+			_ => None,
+		})
+	}
+}
+
+// Note that on drop, we don't have the chance to gracefully spin down each of the remaining handles;
+// we just abort them all. Still better than letting them dangle.
+impl<Spawner, Job: JobTrait> Drop for Jobs<Spawner, Job> {
+	fn drop(&mut self) {
+		for job_handle in self.running.values() {
+			job_handle.abort_handle.abort();
+		}
+	}
+}
+
+/// A basic implementation of a subsystem.
+///
+/// This struct is responsible for handling message traffic between
+/// this subsystem and the overseer. It spawns and kills jobs on the
+/// appropriate Overseer messages, and dispatches standard traffic to
+/// the appropriate job the rest of the time.
+pub struct JobManager<Spawner, Context, Job: JobTrait> {
+	spawner: Spawner,
+	run_args: Job::RunArgs,
+	context: std::marker::PhantomData<Context>,
```

Comment:
We can avoid the `PhantomData` here and reduce the amount of generic parameters by scoping the `Context` generic only to the `fn run` method.

Reply:
We need to add it to `fn run`, `fn handle_incoming`, and `fn handle_outgoing`; the trait bounds get complicated. It's simpler to keep it as a type parameter in `JobManager`.